File: app/src/main/java/com/testproject/kaera/ringtestapp/controllers/base/RefWatchingController.java
package com.testproject.kaera.ringtestapp.controllers.base;

import android.os.Bundle;
import android.support.annotation.NonNull;

import com.bluelinelabs.conductor.ControllerChangeHandler;
import com.bluelinelabs.conductor.ControllerChangeType;

public abstract class RefWatchingController extends ButterKnifeController {

    private boolean hasExited;

    public RefWatchingController() {
        super();
    }

    public RefWatchingController(Bundle args) {
        super(args);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (hasExited) RingApplication.refWatcher.watch(this);
    }

    @Override
    protected void onChangeEnded(@NonNull ControllerChangeHandler changeHandler, @NonNull ControllerChangeType changeType) {
        super.onChangeEnded(changeHandler, changeType);
        hasExited = !changeType.isEnter;
        if (isDestroyed()) RingApplication.refWatcher.watch(this);
    }
}

File: app/src/main/java/com/testproject/kaera/ringtestapp/controllers/base/RefWatchingController.kt
package com.testproject.kaera.ringtestapp.controllers.base

import android.os.Bundle
import com.bluelinelabs.conductor.Controller
import com.bluelinelabs.conductor.ControllerChangeHandler
import com.bluelinelabs.conductor.ControllerChangeType
import com.testproject.kaera.ringtestapp.util.getRefWatcher

abstract class RefWatchingController(bundle: Bundle?) : Controller(bundle) {

    var hasExited: Boolean = false

    override fun onDestroy() {
        super.onDestroy()
        if (hasExited) applicationContext?.getRefWatcher()?.watch(this)
    }

    override fun onChangeEnded(changeHandler: ControllerChangeHandler, changeType: ControllerChangeType) {
        super.onChangeEnded(changeHandler, changeType)
        hasExited = !changeType.isEnter
        if (isDestroyed) applicationContext?.getRefWatcher()?.watch(this)
    }
}